app/services/trainees/create_from_dttp.rb

```diff
@@ -0,0 +1,333 @@
+# frozen_string_literal: true
+
+module Trainees
+  class CreateFromDttp
+    include ServicePattern
+
+    class UnrecognisedStatusError < StandardError; end
+
+    def initialize(dttp_trainee:)
+      @dttp_trainee = dttp_trainee
+      @trainee = Trainee.new(mapped_attributes)
+    end
+
+    def call
+      return if dttp_trainee.provider.blank?
+      return if latest_placement_assignment.blank?
+
+      if training_route.blank?
+        dttp_trainee.non_processable_missing_route!
+        return
+      end
+
+      if training_route == TRAINING_ROUTE_ENUMS[:hpitt_postgrad]
+        dttp_trainee.non_processable_hpitt!
+        return
+      end
+
+      if trainee_already_exists?
+        dttp_trainee.non_processable_duplicate!
+        return
+      end
+
+      trainee.save!
+      create_degrees!
+
+      # TODO: TBC what states are we going to use
+      dttp_trainee.processed!
+
+      trainee
+    end
+
+  private
+
+    attr_reader :dttp_trainee, :trainee
+
+    def mapped_attributes
+      return if latest_placement_assignment.blank?
+
+      {
+        state: trainee_status,
+        provider: dttp_trainee.provider,
+        trainee_id: trainee_id,
+        training_route: training_route,
+        trn: dttp_trainee.response["dfe_trn"],
+        submitted_for_trn_at: latest_placement_assignment.response["dfe_trnassessmentdate"],
+        dttp_id: dttp_trainee.dttp_id,
+        placement_assignment_dttp_id: latest_placement_assignment.dttp_id,
+      }.merge(personal_details_attributes)
+       .merge(contact_attributes)
+       .merge(ethnicity_and_disability_attributes)
+       .merge(course_attributes)
+       .merge(school_attributes)
+       .merge(funding_attributes)
+    end
+
+    def create_degrees!
+      ::Degrees::CreateFromDttp.call(trainee: trainee)
+    end
+
+    def earliest_placement_assignment
+      @earliest_placement_assignment ||= sorted_placement_assignments.first
+    end
+
+    def latest_placement_assignment
+      @latest_placement_assignment ||= sorted_placement_assignments.last
+    end
+
+    def sorted_placement_assignments
+      @sorted_placement_assignments ||= processable_placement_assignments.sort_by do |placement_assignment|
+        placement_assignment.response["dfe_programmestartdate"]
```

Comment:
Can be simplified now we're spun out `programme_start_date` into a separate column.

Some instances below too.

Reply:
Fixed in eaf8aabf5dfc4f1b5bbb5a4cc7143d2db0386fb9